Cleanup remote tracebacks

j-bennet opened the pull request:
Closes https://github.com/dask/distributed/issues/4880.

This adds a setting called `distributed.admin.shorten-traceback`. If `true`, `sys.excepthook` is wrapped, and for all unhandled exceptions, stack traces are shortened to 3 frames max. For `IPython`, `InteractiveShell.showtraceback` is wrapped. In addition, information about worker and task is tucked onto exception `args`, so it looks like this:

```
ValueError: Bad block id (2,)
in task: ('throws_map_blocks-ea77cec762bc72be73248e7d1ca88e18', 2)
on worker: {'tcp://127.0.0.1:61285'}
```

Not quite as pretty as suggested in https://github.com/dask/distributed/issues/4880, but hopefully informative.

Executing this snippet in console:

```python
import dask
import dask.array as da
from distributed import Client


def throws_map_blocks(x, block_id=None):
    if block_id is not None and block_id[0] % 2 == 0:
        raise ValueError(f"Bad block id {block_id}")
        return x / 2


if __name__ == "__main__":
    client = Client(silence_logs="ERROR")
    arr = da.arange(30, chunks=3)
    op = arr.map_blocks(throws_map_blocks)
    op.compute()
```

Before:

<details>
<summary>Click me!</summary>
<pre>
Traceback (most recent call last):
  File "/Users/jbennet/src/distributed/test.py", line 16, in <module>
    op.compute()
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/base.py", line 314, in compute
    (result,) = compute(self, traverse=False, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/base.py", line 599, in compute
    results = schedule(dsk, keys, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/client.py", line 3270, in get
    raise exc_value.with_traceback(exc_traceback)
  File "/Users/jbennet/src/distributed/distributed/client.py", line 3267, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/client.py", line 2402, in gather
    return self.sync(
           ^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/utils.py", line 351, in sync
    return sync(
           ^^^^^
  File "/Users/jbennet/src/distributed/distributed/utils.py", line 418, in sync
    raise exc.with_traceback(tb)
  File "/Users/jbennet/src/distributed/distributed/utils.py", line 391, in f
    result = yield future
             ^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/tornado/gen.py", line 767, in run
    value = future.result()
            ^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/distributed/client.py", line 2265, in _gather
    raise exception.with_traceback(traceback)
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/optimization.py", line 990, in __call__
    return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))
  ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/core.py", line 149, in get
    result = _execute_task(task, cache)
  ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/mambaforge/envs/distributed/lib/python3.11/site-packages/dask/array/core.py", line 523, in _pass_extra_kwargs
    return func(*args[len(keys) :], **kwargs)
      ^^^^^^^^^^^^^^^^^
  File "/Users/jbennet/src/distributed/test.py", line 8, in throws_map_blocks
    raise ValueError(f"Bad block id {block_id}")
  ^^^^^^^^^^^^^^^^^
ValueError: Bad block id (4,)
</pre>
</details>

After:

<details>
<summary>Click me!</summary>
<pre>
Traceback (most recent call last):
  File "/Users/jbennet/src/distributed/test.py", line 17, in <module>
    op.compute()
  File "/Users/jbennet/src/distributed/distributed/client.py", line 3287, in get
    raise exc_value.with_traceback(exc_traceback)
  File "/Users/jbennet/src/distributed/test.py", line 8, in throws_map_blocks
    raise ValueError(f"Bad block id {block_id}")
  ^^^^^^^^^^^^^^^^^
ValueError: Bad block id (2,)
in task: ('throws_map_blocks-ea77cec762bc72be73248e7d1ca88e18', 2)
on worker: {'tcp://127.0.0.1:61285'}
</pre>
</details>

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

GPUtester commented:
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

j-bennet commented:
@fjetter I'd appreciate if you could take a look at this. In #4880, it is also requested that we add worker name and task info to the exception output. In `distributed.Client`, I'm not sure where to get this information.

fjetter commented:
> it is also requested that we add worker name and task info to the exception output. In distributed.Client, I'm not sure where to get this information.

You shouldn't consider what's in this ticket as a request but rather as a suggestion. If we don't have this information easily available, this should not drive implementation effort

j-bennet commented:
@fjetter 

> You shouldn't consider what's in this ticket as a request but rather as a suggestion. If we don't have this information easily available, this should not drive implementation effort

Your response in Slack was helpful with [this part](https://github.com/dask/distributed/pull/7847/commits/edaef8f7bdc0f60e8d4cfefeb72da5d72256205b). I'd rather not override the built-in formatting method to print a custom header on top of an exception, and I don't want to wrap it with a custom Exception, or attach a made-up context, so the easiest thing to do was tucking it into exception `args`, even though this is not quite as pretty.

See what you think.

milesgranger commented on distributed/client.py:
```diff
@@ -130,6 +131,46 @@
+    )
+
+
+def _restore_exception_hook():
+    sys.excepthook = original_excepthook
+
```
Is it needed to restore the original `InteractiveShell.showtraceback` as well then?

j-bennet replied:
Good point, added.

milesgranger reviewed (approved):
Mostly looks good to me, with one clarifying comment. :)

crusaderky reviewed (changes requested):
Could you revisit the tests please?

- `test_client.py::test_truncate_traceback` doesn't add anything to what's already tested by `test_utils.py::test_truncate_traceback`
- you're not testing anywhere that your new function is actually called by client.py (and in two separate use cases)
- you're not testing anywhere that the config flag is actually meaningful

Could you revisit the documentation in the schema? The setting doesn't - I quote - *Truncate traceback to only relevant frames*; that wording leaves me with the impression that what's doing is to remove visibility of dask internals, which is something that's univerally desirable. Keeping the first and last frame is an important detail.

Finally, I'm not sold on the word "truncate" - to me, it strongly evokes removing either the head or the tail from something, but here you're removing *from the middle*.

j-bennet commented:
@crusaderky 

> Finally, I'm not sold on the word "truncate" - to me, it strongly evokes removing either the head or the tail from something, but here you're removing from the middle.

I did `truncate_traceback` because we already have `truncate_exception`, but yes, that one preserves head. I was considering `shorten_traceback` or `cleanup_traceback` as well. Do you have any suggestions?

j-bennet commented:
@crusaderky

All good points, I addressed them as best I could.

* renamed the setting and related function to `shorten-traceback`
* the test called `test_shorten_traceback_excepthook` is testing that excepthook works as expected with different values of `distributed.admin.shorten-traceback`. I was unable to mock `sys.excepthook`, since pytest adds its own excepthook. I reverted to using `popen`. Not sure is there's a better way?
* added a test for ipython handler, called `test_shorten_traceback_showtraceback`.

I'd appreciate if you take another look.

j-bennet commented:
One of the test failures looks legitimate:

```
FAILED distributed/tests/test_utils.py::test_shorten_traceback_showtraceback[True-2] - assert 0 == 2
 +  where 0 = len([])
```

Looking into.

j-bennet commented:
@crusaderky This PR was reworked, and now the only thing it does is pushing worker info into remote exception. Actual traceback cleanup is done in https://github.com/dask/dask/pull/10354. Please take a look. It seems to me that test failures are not related.

crusaderky commented on distributed/client.py:
```diff
@@ -602,9 +602,28 @@ def set_error(self, exception, traceback):
         _, exception, traceback = clean_exception(exception, traceback)
 
+        if (
+            dask.config.get("admin.traceback.shorten", False) is True
```
As a rule of thumb, it's a very bad idea to pass a default to dask.config.get.
With my changes in dask/dask, this test is always false. Could you amend this PR to match?

j-bennet replied:
Updated. About the default value, can you clarify, if the setting is added in `dask`, and not in `distributed`, how can I avoid using the default? Or should I add the same setting on both sides?

crusaderky replied:
Note: every version of distributed pins the matching version of dask, and vice versa.

crusaderky commented on distributed/client.py:
```diff
@@ -602,9 +602,29 @@ def set_error(self, exception, traceback):
         _, exception, traceback = clean_exception(exception, traceback)
 
+        if (
+            dask.config.get("admin.traceback.shorten.when", None)
+            and dask.config.get("admin.traceback.shorten.what", None)
```
```suggestion
            dask.config.get("admin.traceback.shorten.when")
            and dask.config.get("admin.traceback.shorten.what")
```

crusaderky commented on distributed/client.py:
```diff
@@ -602,9 +602,29 @@ def set_error(self, exception, traceback):
         _, exception, traceback = clean_exception(exception, traceback)
 
+        if (
+            dask.config.get("admin.traceback.shorten.when", None)
+            and dask.config.get("admin.traceback.shorten.what", None)
+            and key
+            and erred_on
```
```suggestion
            key
            and erred_on
            and dask.config.get("admin.traceback.shorten.when")
            and dask.config.get("admin.traceback.shorten.what")
```

j-bennet replied:
I'm not disabling worker information. I'm disabling pushing worker information into exception's args.

As you correctly noted, what's going on below is hacky. I want this work with any `Exception`, and without overriding the built-in exception formatter, whether it's python console, ipython console, or ipython notebook. Whatever args were previously attached to the Exception, this will stringify the first arg, and append the worker info to it. This way, the worker info becomes an integral part of the exception, it doesn't have to be a custom exception, and it doesn't add another wrapping layer.

crusaderky commented on distributed/client.py:
```diff
@@ -602,9 +602,29 @@ def set_error(self, exception, traceback):
+                exception.args = (worker_message,)
+            else:
+                arg0 = f"{exception.args[0]}\n{worker_message}"
+                exception.args = (arg0,) + exception.args[1:]
```
What kind of exception is this? is this a *specific* exception type raised by distributed?
args can contain very different things depending on the exception subclass and this change will break many custom exceptions.

crusaderky replied:
This comment has not been addressed.
In fact, I can very easily make it break:

```python
class BadNumbers(Exception):
    def __str__(self):
        return "I didn't like these numbers: " + " ".join(f"{x:.1f}" for x in self.args)

def f():
    raise BadNumbers(1, 2.5, 3)
    
f()

---------------------------------------------------------------------------
BadNumbers                                Traceback (most recent call last)
Cell In[10], line 8
      5 def f():
      6     raise BadNumbers(1, 2.5, 3)
----> 8 f()

Cell In[10], line 6, in f()
      5 def f():
----> 6     raise BadNumbers(1, 2.5, 3)

BadNumbers: I didn't like these numbers: 1.0 2.5 3.0

import distributed
client = distributed.Client()
client.submit(f).result()

Custom TB Handler failed, unregistering
---------------------------------------------------------------------------
BadNumbers                                Traceback (most recent call last)
    [... skipping hidden 1 frame]

Cell In[12], line 1
----> 1 c.submit(f).result()

Cell In[10], line 6, in f()
      5 def f():
----> 6     raise BadNumbers(1, 2.5, 3)

<class 'str'>: (<class 'ValueError'>, ValueError("Unknown format code 'f' for object of type 'str'"))
[...]
TypeError: structured_traceback() missing 1 required positional argument: 'evalue'
```

j-bennet replied:
Ah. I addressed this comment here:

https://github.com/dask/distributed/pull/7847#discussion_r1237787343

But I changed the implementation since, so it's not relevant anymore.

crusaderky commented on distributed/client.py:
```diff
@@ -1618,10 +1638,12 @@ def _handle_retried_key(self, key=None):
 
-    def _handle_task_erred(self, key=None, exception=None, traceback=None):
+    def _handle_task_erred(
+        self, key=None, exception=None, traceback=None, erred_on=None
```
Note: these defaults are there for historical reasons and very frequently you never have a use case for them.

crusaderky reviewed (changes requested):
Please add unit tests

crusaderky commented:
As it is right now, I do not think this PR is a good idea, at all.
Maybe instead of tampering with exception args we could just call logging.error() when the client re-raises?

crusaderky commented:
Re-raising a wrapper exception would make the traceback less readable and break backwards compatibility.

>  but that still means any unhandled exception would not have worker info. 

I don't follow. You'd just call logger.error from Future.result() or whatever re-raises the exception.

j-bennet commented:
@crusaderky

> Re-raising a wrapper exception would make the traceback less readable and break backwards compatibility.

I agree. I tried another way, and it might actually solve that problem. It solves the problem with `BadNumbers` error. See what you think.

j-bennet commented:
@crusaderky

> I don't follow. You'd just call logger.error from Future.result() or whatever re-raises the exception.

I created a branch here to illustrate this approach:

https://github.com/dask/distributed/pull/7947

It's a valid approach, less hacky than the current PR, but user still needs to scroll up for worker info. This is what I meant when I said "but that still means any unhandled exception would not have worker info". The clients logs worker info, but the exception thrown on the client doesn't have it:

```python
---------------------------------------------------------------------------
BadNumbers                                Traceback (most recent call last)
Cell In[5], line 1
----> 1 client.submit(f).result()

Cell In[3], line 2, in f()
      1 def f():
----> 2     raise BadNumbers(1, 2.5, 3)

BadNumbers: I didn't like these numbers: 1.0 2.5 3.0
```

Maybe this is fine. I can go with https://github.com/dask/distributed/pull/7947 too.

j-bennet commented:
Closing in favor of https://github.com/dask/distributed/pull/7947.
